src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeFixes/AbstractDocumentBasedFixAllProvider.cs

```diff
@@ -0,0 +1,232 @@
+            var progressTracker = fixAllContext.GetProgressTracker();
+            progressTracker.Description = this.GetFixAllTitle(fixAllContext);
+
+            var solution = fixAllContext.Solution;
+            progressTracker.AddItems(projectIds.Length);
+
+            using var _1 = PooledDictionary<DocumentId, SyntaxNode>.GetInstance(out var docIdToNewRoot);
+            using var _2 = PooledDictionary<DocumentId, SyntaxNode>.GetInstance(out var docIdToCleanedRoot);
+            using var _3 = ArrayBuilder<Task<(DocumentId docId, SourceText sourceText)>>.GetInstance(out var cleanupTasks);
+
+            var currentSolution = solution;
+            foreach (var projectId in projectIds)
+            {
+                // Clear out the temporary state.
+                docIdToNewRoot.Clear();
+                docIdToCleanedRoot.Clear();
+                cleanupTasks.Clear();
+
+                try
+                {
+                    var project = solution.GetRequiredProject(projectId);
+
+                    // First, determine what the new syntax tree should be for all these documents.
+                    await AddDocumentFixesAsync(fixAllContext, project, docIdToNewRoot).ConfigureAwait(false);
+                    if (docIdToNewRoot.Count == 0)
+                        continue;
+
+                    // Next, go and insert those all into the solution so all the docs in this particular project point
+                    // at the new trees.  At this point though, the trees have not been postprocessed/cleaned.
+                    foreach (var (docId, newRoot) in docIdToNewRoot)
+                        currentSolution = currentSolution.WithDocumentSyntaxRoot(docId, newRoot);
+
+                    // Next, go and cleanup the trees we inserted.  We do this in bulk so we can benefit from Sharing a
+                    // single compilation across all of them for all the semantic work we need to do. 
+                    //
+                    // Also, once we clean the document, get the text of it and insert that back into the final
+                    // solution.  This way we can release both the original fixed tree, and the cleaned tree (both of
+                    // which can be much more expensive than just text).
+
+                    foreach (var (docId, _) in docIdToNewRoot)
+                    {
+                        var dirtyDocument = currentSolution.GetRequiredDocument(docId);
+                        cleanupTasks.Add(Task.Run(async () =>
+                        {
+                            var cleanedDocument = await PostProcessCodeAction.Instance.PostProcessChangesAsync(dirtyDocument, cancellationToken).ConfigureAwait(false);
+                            var cleanedText = await cleanedDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                            return (dirtyDocument.Id, cleanedText);
+                        }));
```

Comment:
note: we still try to parallelize when we can.  but only at the Project level.  Trying to do everyhing at the solution level is just too intensive and often leads to OOMing, and making very little actual progress.  Projects are a nice unit of granularity in general